src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+package com.fulcrumgenomics.fastq
+
+import java.io.File
+import java.nio.file.Path
+import java.util.NoSuchElementException
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadStructure.SubRead
+import com.fulcrumgenomics.util._
+import com.fulcrumgenomics.util.miseq.{Sample, SampleBarcode, SampleSheet}
+import dagr.commons.CommonsDef.{DirPath, FilePath, PathToFastq, _}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.fastq.FastqReader
+import htsjdk.samtools.util._
+
+import scala.collection.mutable
+
+object DemuxFastqs {
+  import SampleBarcode.SampleBarcodeDelimiter
+
+  val UnmatchedSampleId: String = "unmatched"
+
+  /** Custom columns in the Illumina Experiment Manager Sample Sheet. */
+  object CustomSampleSheetHeaderNames {
+    val R1SampleBarcodeBases = "R1_Barcode_Bases"
+    val R2SampleBarcodeBases = "R2_Barcode_Bases"
+  }
+
+  /** Creates the sample output BAM for the given sample.  [setSampleBarcode] should be called first. */
+  def sampleOutputBam(output: Path, sample: Sample): Path = {
```

Comment:
I realize this (and other methods) don't need access to any state, but I think it's nicer to have them inside the class instead of the object for two reasons: 1) it keeps this together a little more, 2) it makes it much easier to extend DemuxFastqs and modify behaviour a little bit (e.g. introduce _another_ custom column).